shared/src/ledger/pos/vp.rs

```diff
@@ -0,0 +1,334 @@
+                let slashes = self
+                    .ctx
+                    .read_pre(&validator_slashes_key(&bond_id.validator))?
+                    .and_then(|bytes| Slashes::try_from_slice(&bytes[..]).ok())
+                    .unwrap_or_default();
+                changes.push(Bond {
+                    id: bond_id.clone(),
+                    data: Data { pre, post },
+                    slashes,
+                });
+            } else if let Some(unbond_id) = is_unbond_key(key) {
+                let pre = self
+                    .ctx
+                    .read_pre(key)?
+                    .and_then(|bytes| Unbonds::try_from_slice(&bytes[..]).ok());
+                let post = self
+                    .ctx
+                    .read_post(key)?
+                    .and_then(|bytes| Unbonds::try_from_slice(&bytes[..]).ok());
+                // For unbonds, we need to look-up slashes
+                let slashes = self
+                    .ctx
+                    .read_pre(&validator_slashes_key(&unbond_id.validator))?
+                    .and_then(|bytes| Slashes::try_from_slice(&bytes[..]).ok())
+                    .unwrap_or_default();
+                changes.push(Unbond {
+                    id: unbond_id.clone(),
+                    data: Data { pre, post },
+                    slashes,
+                });
+            } else if is_total_voting_power_key(key) {
+                let pre = self.ctx.read_pre(key)?.and_then(|bytes| {
+                    TotalVotingPowers::try_from_slice(&bytes[..]).ok()
+                });
+                let post = self.ctx.read_post(key)?.and_then(|bytes| {
+                    TotalVotingPowers::try_from_slice(&bytes[..]).ok()
+                });
+                changes.push(TotalVotingPower(Data { pre, post }));
+            } else {
+                tracing::info!("PoS unrecognized key change {} rejected", key);
+                return Ok(false);
+            }
+        }
+
+        let params = self.read_pos_params();
+        let errors = validate(&params, changes, current_epoch);
+        Ok(if errors.is_empty() {
+            true
+        } else {
+            tracing::info!(
+                "PoS validation errors:\n - {}",
+                errors.iter().format("\n - ")
+            );
+            false
+        })
+    }
+}
+
+impl<D, H> PosReadOnly for PosVP<'_, D, H>
+where
+    D: 'static + ledger_storage::DB + for<'iter> ledger_storage::DBIter<'iter>,
+    H: 'static + StorageHasher,
+{
+    type Address = Address;
+    type PublicKey = key::ed25519::PublicKey;
+    type TokenAmount = token::Amount;
+    type TokenChange = token::Change;
+
+    const POS_ADDRESS: Self::Address = super::ADDRESS;
+
+    fn staking_token_address() -> Self::Address {
+        super::staking_token_address()
+    }
+
+    fn read_pos_params(&self) -> PosParams {
+        let value = self.ctx.read_pre(&params_key()).unwrap().unwrap();
```

Comment:
just a question.
Why does `PosReadOnly` for `PosVP` always read the prior states?

Reply:
This is because the trait interface in the PoS crate doesn't fit very nicely with the integration in here (The `PosReadOnly` is also used in transactions integration, which only need to read the current state, not pre/post). We have this follow-up issue to make this nicer https://github.com/anoma/anoma/issues/443